src/OpenTelemetry.Api/Trace/SpanAttributes.cs

```diff
@@ -43,9 +43,13 @@ public SpanAttributes(IEnumerable<KeyValuePair<string, object>> attributes)
         {
             Guard.ThrowIfNull(attributes);
 
+            var attributeCountLimit = TracerProvider.SpanAttributeCountLimit ?? -1;
             foreach (KeyValuePair<string, object> kvp in attributes)
             {
-                this.AddInternal(kvp.Key, kvp.Value);
+                if (attributeCountLimit-- != 0)
+                {
+                    this.AddInternal(kvp.Key, kvp.Value);
+                }
```

Comment:
Need to exit the loop once the limit is reached (or it will underflow to negative)

Reply:
If negative then it will loop over all attributes and then stop. `attributeCountLimit` is initialized to -1 for this purpose when the user set `TracerProvider.SpanAttributeCountLimit = null`.

Though, for efficiency, this loop could break when the limit has been reached. Also will add tests to demonstrate when limits are not set.